TREnvironmentEditor/Helpers/EMLocationUtilities.cs

```diff
@@ -0,0 +1,55 @@
+﻿using System.Linq;
+using TRFDControl;
+using TRFDControl.FDEntryTypes;
+using TRFDControl.Utilities;
+using TRLevelReader.Helpers;
+using TRLevelReader.Model;
+using TRLevelReader.Model.Enums;
+
+namespace TREnvironmentEditor.Helpers
+{
+    public static class EMLocationUtilities
+    {
+        public static int GetContainedSecret(this EMLocation location, TRLevel level, FDControl floorData)
+        {
+            TRRoomSector sector = FDUtilities.GetRoomSector(location.X, location.Y, location.Z, location.Room, level, floorData);
+            return GetSectorSecret(sector, floorData);
+        }
+
+        public static int GetContainedSecret(this EMLocation location, TR2Level level, FDControl floorData)
+        {
+            TRRoomSector sector = FDUtilities.GetRoomSector(location.X, location.Y, location.Z, location.Room, level, floorData);
+            for (int i = 0; i < level.NumEntities; i++)
+            {
+                TR2Entity entity = level.Entities[i];
+                if (TR2EntityUtilities.IsSecretType((TR2Entities)entity.TypeID)
+                    && FDUtilities.GetRoomSector(location.X, location.Y, location.Z, location.Room, level, floorData) == sector)
+                {
+                    return i;
+                }
+            }
+            return -1;
+        }
+
+        public static int GetContainedSecret(this EMLocation location, TR3Level level, FDControl floorData)
+        {
+            TRRoomSector sector = FDUtilities.GetRoomSector(location.X, location.Y, location.Z, location.Room, level, floorData);
+            return GetSectorSecret(sector, floorData);
+        }
+
+        private static int GetSectorSecret(TRRoomSector sector, FDControl floorData)
+        {
+            if (sector.FDIndex != 0)
+            {
+                if (floorData.Entries[sector.FDIndex].Find(e => e is FDTriggerEntry) is FDTriggerEntry trigger
+                    && trigger.TrigType == FDTrigType.Pickup
+                    && trigger.TrigActionList.Find(a => a.TrigAction == FDTrigAction.SecretFound) != null)
+                {
+                    return trigger.TrigActionList.First().Parameter;
```

Comment:
Should this be taking `.First(a => a.TrigAction == FDTrigAction.SecretFound)`? I don't know about when it's Pickup, but if there's any situations like this trigger in Tinnos it would return 44 for the secret index.
![image](https://user-images.githubusercontent.com/4263940/218225123-ea1b30c4-e43c-4c65-a8a0-f6b0227ca476.png)

Reply:
It's the item index we want to return here rather than the secret index - so for pickup triggers the first object entry in the action list marks the thing being picked up. I've renamed the method to make it clearer and added a check to make sure that the object entry is present instead of relying on `.First()` actually returning something.